tensorflow/contrib/model_average/python/device_setter.py

```diff
@@ -0,0 +1,205 @@
+    """Create a new `_RoundRobinStrategy`.
+
+    Args:
+      num_tasks: Number of ps tasks to cycle among.
+    """
+    self._num_tasks = num_tasks
+    self._next_task = 0
+
+  def __call__(self, unused_op):
+    """Choose a ps task index for the given `Operation`.
+
+    Args:
+      unused_op: An `Operation` to be placed on ps.
+
+    Returns:
+      The next ps task index to use for the `Operation`. Returns the next
+      index, in the range `[offset, offset + num_tasks)`.
+    """
+    task = self._next_task
+    self._next_task = (self._next_task + 1) % self._num_tasks
+    return task
+
+
+class _ReplicaDeviceChooser(object):
+  """Class to choose devices for Ops in a replicated training setup.
+
+  This class is not to be used directly by users.  See instead
+  `model_average_device_setter()` below.
+  """
+
+  def __init__(self, ps_tasks, ps_device, worker_device, merge_devices, ps_ops,
+               ps_strategy):
+    """Create a new `_ReplicaDeviceChooser`.
+
+    Args:
+      ps_tasks: Number of tasks in the `ps` job.
+      ps_device: String.  Name of the `ps` job.
+      worker_device: String.  Name of the `worker` job.
+      merge_devices: Boolean. Set to True to allow merging of device specs.
+      ps_ops: List of strings representing `Operation` types that need to be
+        placed on `ps` devices.
+      ps_strategy: A callable invoked for every ps `Operation` (i.e. matched by
+        `ps_ops`), that takes the `Operation` and returns the ps task index to
+        use.
+    """
+    self._ps_tasks = ps_tasks
+    self._ps_device = ps_device
+    self._worker_device = worker_device
+    self._merge_devices = merge_devices
+    self._ps_ops = ps_ops
+    self._ps_strategy = ps_strategy
+
+  def device_function(self, op):
+    """Choose a device for `op`.
+
+    Args:
+      op: an `Operation`.
+
+    Returns:
+      The device to use for the `Operation`.
+    """
+    if not self._merge_devices and op.device:
+      return op.device
+    current_device = pydev.DeviceSpec.from_string(op.device or "")
+    spec = pydev.DeviceSpec()
+
+    node_def = op if isinstance(op, node_def_pb2.NodeDef) else op.node_def
+    # ma global variables
+    # TODO(chen meng): only global variables should be placed on ps.
+    if node_def.name.find('modelAverage') != -1:
+      device_string = "%s/task:%d" % (
+        self._ps_device, self._ps_strategy(op))
+      if self._merge_devices:
+        spec = pydev.DeviceSpec.from_string(device_string)
+        spec.merge_from(current_device)
+        return spec.to_string()
+      else:
+        return device_string
+    else:
+      if self._worker_device:
+        if not self._merge_devices:
+          return self._worker_device
+        spec = pydev.DeviceSpec.from_string(self._worker_device)
+
+      if not self._merge_devices:
+        return ""
+
+    spec.merge_from(current_device)
+    return spec.to_string()
+
+
+def model_average_device_setter(ps_tasks=0, ps_device="/job:ps",
```

Comment:
There is only one difference:
tf.train.replica_device_setter placed all variables(including global/local variables) on ps, while in model average, each worker own its o local variables （local model parameters），these local variables should be placed in each worker.